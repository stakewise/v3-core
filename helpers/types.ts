import { BigNumberish } from 'ethers'
export type ThenArg<T> = T extends PromiseLike<infer U> ? U : T

export enum Networks {
  mainnet = 'mainnet',
  goerli = 'goerli',
}

export type NetworkConfig = {
  url: string
  chainId: number

  governor: string
  validatorsRegistry: string
  securityDeposit: BigNumberish

  // Keeper
  oracles: string[]
  rewardsMinOracles: number
  validatorsMinOracles: number
  rewardsDelay: number
  oraclesConfigIpfsHash: string
  maxAvgRewardPerSecond: BigNumberish

  // OsToken
  treasury: string
  osTokenFeePercent: number
  osTokenCapacity: BigNumberish
  osTokenName: string
  osTokenSymbol: string

  // OsTokenConfig
  redeemFromLtvPercent: number
  redeemToLtvPercent: number
  liqThresholdPercent: number
  liqBonusPercent: number
  ltvPercent: number

  // GenesisEthVault
  genesisVault: {
    admin: string
    poolEscrow: string
    stakedEthToken: string
    capacity: BigNumberish
    feePercent: number
  }
}
