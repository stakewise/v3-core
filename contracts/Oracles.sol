// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.17;

import {Ownable} from '@openzeppelin/contracts/access/Ownable.sol';
import {EIP712} from '@openzeppelin/contracts/utils/cryptography/EIP712.sol';
import {ECDSA} from '@openzeppelin/contracts/utils/cryptography/ECDSA.sol';
import {IOracles} from './interfaces/IOracles.sol';

/**
 * @title Oracles
 * @author StakeWise
 * @notice Defines the functionality for verifying signatures of the whitelisted off-chain oracles
 */
contract Oracles is Ownable, EIP712, IOracles {
  /// @inheritdoc IOracles
  mapping(address => bool) public override isOracle;

  /// @inheritdoc IOracles
  uint256 public override totalOracles;

  /// @inheritdoc IOracles
  uint256 public override requiredOracles;

  /**
   * @dev Constructor
   * @param _owner The address of the contract owner
   * @param initialOracles The addresses of the initial oracles
   * @param initialRequiredOracles The number or required oracles for the verification
   */
  constructor(
    address _owner,
    address[] memory initialOracles,
    uint256 initialRequiredOracles
  ) EIP712('Oracles', '1') {
    for (uint256 i = 0; i < initialOracles.length; ) {
      addOracle(initialOracles[i]);
      unchecked {
        ++i;
      }
    }
    setRequiredOracles(initialRequiredOracles);
    _transferOwnership(_owner);
  }

  /// @inheritdoc IOracles
  function addOracle(address oracle) public override onlyOwner {
    if (isOracle[oracle]) revert AlreadyAdded();

    isOracle[oracle] = true;
    unchecked {
      // cannot realistically overflow
      totalOracles += 1;
    }
    emit OracleAdded(oracle);
  }

  /// @inheritdoc IOracles
  function removeOracle(address oracle) external override onlyOwner {
    if (!isOracle[oracle]) revert AlreadyRemoved();

    // SLOAD to memory
    uint256 _totalOracles;
    unchecked {
      // cannot underflow
      _totalOracles = totalOracles - 1;
    }

    isOracle[oracle] = false;
    totalOracles = _totalOracles;
    if (_totalOracles < requiredOracles) setRequiredOracles(_totalOracles);

    emit OracleRemoved(oracle);
  }

  /// @inheritdoc IOracles
  function setRequiredOracles(uint256 _requiredOracles) public override onlyOwner {
    if (_requiredOracles == 0 || totalOracles < _requiredOracles) revert InvalidRequiredOracles();
    requiredOracles = _requiredOracles;
    emit RequiredOraclesUpdated(_requiredOracles);
  }

  /// @inheritdoc IOracles
  function verifySignatures(bytes32 message, bytes calldata signatures) external view override {
    // SLOAD to memory
    uint256 _requiredOracles = requiredOracles;

    // check whether enough signatures
    if (signatures.length < _requiredOracles * 65) revert NotEnoughSignatures();

    bytes32 data = _hashTypedDataV4(message);
    address lastOracle;
    address currentOracle;
    uint256 startIndex;
    for (uint256 i = 0; i < _requiredOracles; ) {
      unchecked {
        // cannot overflow as signatures.length is checked above
        currentOracle = ECDSA.recover(data, signatures[startIndex:startIndex + 65]);
      }
      if (currentOracle <= lastOracle || !isOracle[currentOracle]) revert InvalidOracle();

      // update last oracle
      lastOracle = currentOracle;

      unchecked {
        // cannot overflow as it's capped with requiredOracles
        ++i;
        startIndex += 65;
      }
    }
  }
}
