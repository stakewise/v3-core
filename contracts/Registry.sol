// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.17;

import {Ownable} from '@openzeppelin/contracts/access/Ownable.sol';
import {IRegistry} from './interfaces/IRegistry.sol';

/**
 * @title Registry
 * @author StakeWise
 * @notice Defines the registry functionality that keeps track of vaults, factories and their upgrades
 */
contract Registry is Ownable, IRegistry {
  /// @inheritdoc IRegistry
  mapping(address => bool) public override vaults;

  /// @inheritdoc IRegistry
  mapping(address => bool) public override factories;

  /// @inheritdoc IRegistry
  mapping(address => address) public override upgrades;

  /**
   * @dev Constructor
   * @param owner_ The address of the registry owner
   */
  constructor(address owner_) {
    _transferOwnership(owner_);
  }

  /// @inheritdoc IRegistry
  function addVault(address vault) external override {
    if (!factories[msg.sender]) revert AccessDenied();
    vaults[vault] = true;
    emit VaultAdded(msg.sender, vault);
  }

  /// @inheritdoc IRegistry
  function addUpgrade(address prevImpl, address newImpl) external override onlyOwner {
    if (prevImpl == newImpl) revert InvalidUpgrade();
    upgrades[prevImpl] = newImpl;
    emit UpgradeAdded(prevImpl, newImpl);
  }

  /// @inheritdoc IRegistry
  function addFactory(address factory) external override onlyOwner {
    if (factories[factory]) revert AlreadyAdded();
    factories[factory] = true;
    emit FactoryAdded(factory);
  }

  /// @inheritdoc IRegistry
  function removeFactory(address factory) external override onlyOwner {
    if (!factories[factory]) revert AlreadyRemoved();
    factories[factory] = false;
    emit FactoryRemoved(factory);
  }
}
