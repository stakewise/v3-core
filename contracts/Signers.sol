// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.17;

import {Ownable} from '@openzeppelin/contracts/access/Ownable.sol';
import {EIP712} from '@openzeppelin/contracts/utils/cryptography/EIP712.sol';
import {ECDSA} from '@openzeppelin/contracts/utils/cryptography/ECDSA.sol';
import {ISigners} from './interfaces/ISigners.sol';

/**
 * @title Signers
 * @author StakeWise
 * @notice Defines the functionality for verifying signatures of the whitelisted off-chain signers
 */
contract Signers is Ownable, EIP712, ISigners {
  /// @inheritdoc ISigners
  mapping(address => bool) public override isSigner;

  /// @inheritdoc ISigners
  uint256 public override totalSigners;

  /// @inheritdoc ISigners
  uint256 public override requiredSigners;

  /**
   * @dev Constructor
   * @param _owner The address of the contract owner
   * @param initialSigners The addresses of the initial signers
   * @param initialRequiredSigners The number or required signers for the verification
   */
  constructor(
    address _owner,
    address[] memory initialSigners,
    uint256 initialRequiredSigners
  ) EIP712('Signers', '1') {
    for (uint256 i = 0; i < initialSigners.length; ) {
      addSigner(initialSigners[i]);
      unchecked {
        ++i;
      }
    }
    setRequiredSigners(initialRequiredSigners);
    _transferOwnership(_owner);
  }

  /// @inheritdoc ISigners
  function addSigner(address signer) public override onlyOwner {
    if (isSigner[signer]) revert AlreadyAdded();

    isSigner[signer] = true;
    unchecked {
      // cannot realistically overflow
      totalSigners += 1;
    }
    emit SignerAdded(signer);
  }

  /// @inheritdoc ISigners
  function removeSigner(address signer) external override onlyOwner {
    if (!isSigner[signer]) revert AlreadyRemoved();

    // SLOAD to memory
    uint256 _totalSigners;
    unchecked {
      // cannot underflow
      _totalSigners = totalSigners - 1;
    }

    isSigner[signer] = false;
    totalSigners = _totalSigners;
    if (_totalSigners < requiredSigners) setRequiredSigners(_totalSigners);

    emit SignerRemoved(signer);
  }

  /// @inheritdoc ISigners
  function setRequiredSigners(uint256 _requiredSigners) public override onlyOwner {
    if (_requiredSigners == 0 || totalSigners < _requiredSigners) revert InvalidRequiredSigners();
    requiredSigners = _requiredSigners;
    emit RequiredSignersUpdated(_requiredSigners);
  }

  /// @inheritdoc ISigners
  function verifySignatures(bytes32 message, bytes calldata signatures) external view override {
    // SLOAD to memory
    uint256 _requiredSigners = requiredSigners;

    // check whether enough signatures
    if (signatures.length < _requiredSigners * 65) revert NotEnoughSignatures();

    bytes32 data = _hashTypedDataV4(message);
    address lastSigner;
    address currentSigner;
    uint256 startIndex;
    for (uint256 i = 0; i < _requiredSigners; ) {
      unchecked {
        // cannot overflow as signatures.length is checked above
        currentSigner = ECDSA.recover(data, signatures[startIndex:startIndex + 65]);
      }
      if (currentSigner <= lastSigner || !isSigner[currentSigner]) revert InvalidSigner();

      // update last signer
      lastSigner = currentSigner;

      unchecked {
        // cannot overflow as it's capped with requiredSigners
        ++i;
        startIndex += 65;
      }
    }
  }
}
