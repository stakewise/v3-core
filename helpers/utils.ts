import { Contract } from 'ethers'
import { HardhatRuntimeEnvironment } from 'hardhat/src/types/runtime'

export async function deployContract(tx: any): Promise<Contract> {
  const result = await tx
  await result.deployTransaction.wait()
  return result
}

async function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export async function verify(
  hre: HardhatRuntimeEnvironment,
  address: string,
  constructorArgs: any,
  contractPath: string
) {
  if (!process.env.BLOCK_EXPLORER_KEY) {
    return
  }

  let count = 0
  const maxTries = 8
  // eslint-disable-next-line no-constant-condition
  while (true) {
    await delay(10000)
    try {
      console.log('Verifying contract at', address)
      await hre.run('verify:verify', {
        address,
        constructorArguments: constructorArgs,
        contract: contractPath,
      })
      break
    } catch (error) {
      if (String(error).includes('Already Verified')) {
        console.log(`Already verified contract at ${contractPath} at address ${address}`)
        break
      }
      if (++count == maxTries) {
        console.log(
          `Failed to verify contract at ${contractPath} at address ${address}, error: ${error}`
        )
        break
      }
      console.log(`Retrying... Retry #${count}, last error: ${error}`)
    }
  }
}
