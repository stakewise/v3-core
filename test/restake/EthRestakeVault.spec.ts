import { loadFixture } from '@nomicfoundation/hardhat-toolbox/network-helpers'
import { ethers } from 'hardhat'
import { Contract, parseEther, Signer, Wallet } from 'ethers'
import { EthRestakeVault, Keeper } from '../../typechain-types'
import { ethRestakeVaultFixture } from '../shared/restakeFixtures'
import { ThenArg } from '../../helpers/types'
import {
  MAX_UINT256,
  VALIDATORS_DEADLINE,
  VALIDATORS_MIN_ORACLES,
  ZERO_ADDRESS,
  ZERO_BYTES32,
} from '../shared/constants'
import { expect } from '../shared/expect'
import { extractEigenPodOwner, toHexString } from '../shared/utils'
import {
  createEthValidatorsData,
  EthValidatorsData,
  exitSignatureIpfsHashes,
  getEthValidatorsSigningData,
  getWithdrawalCredentials,
} from '../shared/validators'
import snapshotGasCost from '../shared/snapshotGasCost'
import { getOraclesSignatures } from '../shared/fixtures'
import { UintNumberType } from '@chainsafe/ssz'
import { getEigenPodManager } from '../shared/contracts'
import keccak256 from 'keccak256'

const gwei = 1000000000n
const uintSerializer = new UintNumberType(8)
const validatorDeposit = parseEther('32')

describe('EthRestakeVault', () => {
  const capacity = MAX_UINT256
  const feePercent = 1000
  const metadataIpfsHash = 'bafkreidivzimqfqtoqxkrpge6bjyhlvxqs3rhe73owtmdulaxr5do5in7u'

  let admin: Signer,
    operatorsManager: Wallet,
    validatorsManager: Wallet,
    other: Wallet,
    withdrawalsManager: Wallet
  let vault: EthRestakeVault, keeper: Keeper, validatorsRegistry: Contract
  let createVault: ThenArg<ReturnType<typeof ethRestakeVaultFixture>>['createEthRestakeVault']

  before('create fixture loader', async function () {
    ;[admin, operatorsManager, validatorsManager, withdrawalsManager, other] = await (
      ethers as any
    ).getSigners()
  })

  beforeEach('deploy fixture', async () => {
    ;({
      createEthRestakeVault: createVault,
      keeper,
      validatorsRegistry,
    } = await loadFixture(ethRestakeVaultFixture))
    vault = await createVault(admin, {
      capacity,
      feePercent,
      metadataIpfsHash,
    })
  })

  it('initializes correctly', async () => {
    await expect(vault.connect(admin).initialize(ZERO_BYTES32)).to.revertedWithCustomError(
      vault,
      'InvalidInitialization'
    )
    expect(await vault.capacity()).to.be.eq(capacity)

    // VaultAdmin
    const adminAddr = await admin.getAddress()

    // VaultVersion
    expect(await vault.version()).to.be.eq(2)
    expect(await vault.vaultId()).to.be.eq(`0x${keccak256('EthRestakeVault').toString('hex')}`)

    // VaultFee
    expect(await vault.admin()).to.be.eq(adminAddr)
    expect(await vault.feeRecipient()).to.be.eq(adminAddr)
    expect(await vault.feePercent()).to.be.eq(feePercent)
  })

  describe('create eigen pod', () => {
    let eigenPodManager: Contract

    beforeEach(async () => {
      eigenPodManager = await getEigenPodManager()
    })

    it('fails for non-operators manager', async () => {
      await expect(vault.connect(other).createEigenPod()).to.be.revertedWithCustomError(
        vault,
        'AccessDenied'
      )
    })

    it('succeeds', async () => {
      const receipt = await vault.connect(admin).createEigenPod()
      expect(await vault.getEigenPods()).to.have.lengthOf(1)

      const eigenPodOwner = await extractEigenPodOwner(receipt)
      const eigenPod = (await vault.getEigenPods())[0]
      await expect(receipt).to.emit(vault, 'EigenPodCreated').withArgs(eigenPodOwner, eigenPod)
      expect(await eigenPodManager.ownerToPod(eigenPodOwner)).to.eq(eigenPod)
      await snapshotGasCost(receipt)
    })
  })

  describe('restake operators manager', () => {
    it('defaults to admin', async () => {
      expect(await vault.restakeOperatorsManager()).to.equal(await admin.getAddress())
    })

    it('fails to set by non-admin', async () => {
      await expect(
        vault.connect(other).setRestakeOperatorsManager(other.address)
      ).to.be.revertedWithCustomError(vault, 'AccessDenied')
    })

    it('fails to set to the same address', async () => {
      await vault.connect(admin).setRestakeOperatorsManager(other.address)
      await expect(
        vault.connect(admin).setRestakeOperatorsManager(other.address)
      ).to.be.revertedWithCustomError(vault, 'ValueNotChanged')
    })

    it('succeeds', async () => {
      const receipt = await vault
        .connect(admin)
        .setRestakeOperatorsManager(operatorsManager.address)
      expect(await vault.restakeOperatorsManager()).to.equal(operatorsManager.address)
      await expect(receipt)
        .to.emit(vault, 'RestakeOperatorsManagerUpdated')
        .withArgs(operatorsManager.address)
      await snapshotGasCost(receipt)
    })
  })

  describe('restake withdrawals manager', () => {
    it('defaults to admin', async () => {
      expect(await vault.restakeWithdrawalsManager()).to.equal(await admin.getAddress())
    })

    it('fails to set by non-admin', async () => {
      await expect(
        vault.connect(other).setRestakeWithdrawalsManager(other.address)
      ).to.be.revertedWithCustomError(vault, 'AccessDenied')
    })

    it('fails to set to the same address', async () => {
      await vault.connect(admin).setRestakeWithdrawalsManager(other.address)
      await expect(
        vault.connect(admin).setRestakeWithdrawalsManager(other.address)
      ).to.be.revertedWithCustomError(vault, 'ValueNotChanged')
    })

    it('succeeds', async () => {
      const receipt = await vault
        .connect(admin)
        .setRestakeWithdrawalsManager(withdrawalsManager.address)
      expect(await vault.restakeWithdrawalsManager()).to.equal(withdrawalsManager.address)
      await expect(receipt)
        .to.emit(vault, 'RestakeWithdrawalsManagerUpdated')
        .withArgs(withdrawalsManager.address)
      await snapshotGasCost(receipt)
    })
  })

  describe('single validator', () => {
    const deadline = VALIDATORS_DEADLINE
    let validatorsData: EthValidatorsData
    let eigenPod: string

    beforeEach(async () => {
      await vault.connect(admin).setRestakeOperatorsManager(operatorsManager.address)
      await vault.connect(admin).setValidatorsManager(validatorsManager.address)
      await vault.connect(operatorsManager).createEigenPod()
      await vault.connect(other).deposit(other.address, ZERO_ADDRESS, { value: validatorDeposit })
      eigenPod = (await vault.getEigenPods())[0]
      validatorsData = await createEthValidatorsData(vault)
    })

    it('fails with invalid validator length', async () => {
      let invalidValidator = validatorsData.validators[0]
      invalidValidator = invalidValidator.subarray(0, invalidValidator.length - 1)
      const exitSignaturesIpfsHash = exitSignatureIpfsHashes[0]
      const validatorsRegistryRoot = await validatorsRegistry.get_deposit_root()
      const signatures = getOraclesSignatures(
        await getEthValidatorsSigningData(
          invalidValidator,
          deadline,
          exitSignaturesIpfsHash,
          keeper,
          vault,
          validatorsRegistryRoot
        ),
        VALIDATORS_MIN_ORACLES
      )

      await expect(
        vault.connect(validatorsManager).registerValidators(
          {
            validatorsRegistryRoot,
            validators: invalidValidator,
            deadline,
            signatures,
            exitSignaturesIpfsHash,
          },
          '0x'
        )
      ).to.be.revertedWithCustomError(vault, 'InvalidValidators')
    })

    it('fails with invalid withdrawal address', async () => {
      const invalidValidator = Buffer.concat([
        validatorsData.validators[0].subarray(0, 176),
        ethers.getBytes(await vault.getAddress()),
      ])
      const exitSignaturesIpfsHash = exitSignatureIpfsHashes[0]
      const validatorsRegistryRoot = await validatorsRegistry.get_deposit_root()
      const signatures = getOraclesSignatures(
        await getEthValidatorsSigningData(
          invalidValidator,
          deadline,
          exitSignaturesIpfsHash,
          keeper,
          vault,
          validatorsRegistryRoot
        ),
        VALIDATORS_MIN_ORACLES
      )

      await expect(
        vault.connect(validatorsManager).registerValidators(
          {
            validatorsRegistryRoot,
            validators: invalidValidator,
            deadline,
            signatures,
            exitSignaturesIpfsHash,
          },
          '0x'
        )
      ).to.be.revertedWithCustomError(vault, 'EigenPodNotFound')
    })

    it('succeeds', async () => {
      const index = await validatorsRegistry.get_deposit_count()
      const validatorsRegistryRoot = await validatorsRegistry.get_deposit_root()
      const validator = validatorsData.validators[0]
      const signatures = getOraclesSignatures(
        await getEthValidatorsSigningData(
          validator,
          deadline,
          exitSignatureIpfsHashes[0],
          keeper,
          vault,
          validatorsRegistryRoot
        ),
        VALIDATORS_MIN_ORACLES
      )

      const receipt = await vault.connect(validatorsManager).registerValidators(
        {
          validatorsRegistryRoot,
          validators: validator,
          deadline,
          signatures,
          exitSignaturesIpfsHash: exitSignatureIpfsHashes[0],
        },
        '0x'
      )
      const publicKey = `0x${validator.subarray(0, 48).toString('hex')}`
      await expect(receipt).to.emit(vault, 'ValidatorRegistered').withArgs(publicKey)
      await expect(receipt)
        .to.emit(validatorsRegistry, 'DepositEvent')
        .withArgs(
          publicKey,
          toHexString(getWithdrawalCredentials(eigenPod)),
          toHexString(Buffer.from(uintSerializer.serialize(Number(validatorDeposit / gwei)))),
          toHexString(validator.subarray(48, 144)),
          index
        )
      await snapshotGasCost(receipt)
    })
  })

  describe('multiple validators', () => {
    const deadline = VALIDATORS_DEADLINE
    let validatorsData: EthValidatorsData
    let eigenPod: string

    beforeEach(async () => {
      await vault.connect(admin).setRestakeOperatorsManager(operatorsManager.address)
      await vault.connect(admin).setValidatorsManager(validatorsManager.address)
      await vault.connect(operatorsManager).createEigenPod()
      eigenPod = (await vault.getEigenPods())[0]
      validatorsData = await createEthValidatorsData(vault)
      await vault.connect(other).deposit(other.address, ZERO_ADDRESS, {
        value: validatorDeposit * BigInt(validatorsData.validators.length),
      })
    })

    it('succeeds', async () => {
      const validatorsRegistryRoot = await validatorsRegistry.get_deposit_root()
      const exitSignaturesIpfsHash = exitSignatureIpfsHashes[0]
      const signatures = getOraclesSignatures(
        await getEthValidatorsSigningData(
          Buffer.concat(validatorsData.validators),
          deadline,
          exitSignaturesIpfsHash,
          keeper,
          vault,
          validatorsRegistryRoot
        ),
        VALIDATORS_MIN_ORACLES
      )
      const approvalParams = {
        validatorsRegistryRoot,
        validators: Buffer.concat(validatorsData.validators),
        signatures,
        exitSignaturesIpfsHash,
        deadline,
      }
      const startIndex = uintSerializer.deserialize(
        ethers.getBytes(await validatorsRegistry.get_deposit_count())
      )
      const receipt = await vault
        .connect(validatorsManager)
        .registerValidators(approvalParams, '0x')
      for (let i = 0; i < validatorsData.validators.length; i++) {
        const validator = validatorsData.validators[i]
        const publicKey = toHexString(validator.subarray(0, 48))
        await expect(receipt).to.emit(vault, 'ValidatorRegistered').withArgs(publicKey)
        await expect(receipt)
          .to.emit(validatorsRegistry, 'DepositEvent')
          .withArgs(
            publicKey,
            toHexString(getWithdrawalCredentials(eigenPod)),
            toHexString(Buffer.from(uintSerializer.serialize(Number(validatorDeposit / gwei)))),
            toHexString(validator.subarray(48, 144)),
            toHexString(Buffer.from(uintSerializer.serialize(startIndex + i)))
          )
      }
      await snapshotGasCost(receipt)
    })
  })
})
