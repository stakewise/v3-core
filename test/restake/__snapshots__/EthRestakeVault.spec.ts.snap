// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EthRestakeVault create eigen pod succeeds 1`] = `
Object {
  "calldataByteLength": 4,
  "gasUsed": 651839,
}
`;

exports[`EthRestakeVault multiple validators succeeds 1`] = `
Object {
  "calldataByteLength": 2820,
  "gasUsed": 556020,
}
`;

exports[`EthRestakeVault restake operators manager succeeds 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 52674,
}
`;

exports[`EthRestakeVault restake withdrawals manager succeeds 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 51882,
}
`;

exports[`EthRestakeVault single validator succeeds 1`] = `
Object {
  "calldataByteLength": 1060,
  "gasUsed": 268348,
}
`;
